examples/supply_chain/rl/algorithms/base_stock_policy.py

```diff
@@ -0,0 +1,75 @@
+
+import numpy as np
+import cvxpy as cp
+from .rule_based import RuleBasedPolicy
+from typing import List
+
+class BaseStockPolicy(RuleBasedPolicy):
+    def __init__(self, name: str, policy_para) -> None:
+        super().__init__(name)
+
+        self.update_frequency = policy_para["update_frequency"]
+        self.data_loader = policy_para["data_loader"]
+        self.step = {}
+        self.stock_quantity = {}
+
+    def calculate_stock_quantity(self, input_df, state, current_index):
+        time_hrz_len = len(input_df)
+        price = np.round(input_df["price"], 1)
+        storage_cost = np.round(input_df["storage_cost"], 1)
+        order_cost = np.round(input_df["order_cost"], 1)
+        demand = np.round(input_df["demand"], 1)
+
+        stocks = cp.Variable(time_hrz_len + 1, integer=True)
+        transits = cp.Variable(time_hrz_len + 1, integer=True)
+        sales = cp.Variable(time_hrz_len, integer=True)
+        buy = cp.Variable(time_hrz_len * 2, integer=True)
+        buy_in = cp.Variable(time_hrz_len, integer=True)
+        buy_arv = cp.Variable(time_hrz_len, integer=True)
+        inv_pos = cp.Variable(time_hrz_len, integer=True)
+        target_stock = cp.Variable(time_hrz_len, integer=True)
+        profit = cp.Variable(1)
+        buy_in.value = np.ones(time_hrz_len, dtype = np.int)
+        
+        #add constraints
+        constrs=[]
+        constrs.extend([
+            stocks >= 0, transits >= 0, sales >= 0, buy_in>=0, buy_arv>=0, buy>=0, buy[:time_hrz_len]==0,
+            stocks[current_index] == state["product_level"],
+            transits[current_index] == state["in_transition_quantity"],
+            stocks[1:time_hrz_len+1] == stocks[0:time_hrz_len] + buy_arv - sales,
+            transits[1:time_hrz_len+1] == transits[0:time_hrz_len] - buy_arv + buy_in,
+            sales <= stocks[0:time_hrz_len],
+            sales <= demand,
+            buy_in == buy[time_hrz_len:],
+            inv_pos == stocks[0:time_hrz_len] + transits[0:time_hrz_len],
+            buy_arv == buy[time_hrz_len - state["cur_vlt"] : 2 * time_hrz_len - state["cur_vlt"]],
+            target_stock == inv_pos + buy_in,
+            profit == cp.sum(cp.multiply(price, sales) - cp.multiply(order_cost, buy_in) - cp.multiply(storage_cost, stocks[1:])),
+        ])
+        
+        obj = cp.Maximize(profit)
+        prob = cp.Problem(obj, constrs)
+        prob.solve(solver = cp.GLPK_MI, verbose = True)
+        return target_stock.value
+
+    def _get_action_quantity(self, state: dict) -> int:
+        entity_id = state["entity_id"]
+        if entity_id not in self.step:
+            self.step[entity_id] = 0
+        state["step"] = self.step[entity_id]
```

Comment:
what is `step`? Can we update it in or_agent_state?

Reply:
tick=self._env.tick,